File: user-service/src/main/java/com/example/user/service/AuthService.java
package com.example.user.service;

import com.example.user.client.KakaoApiClient;
import com.example.user.dto.KakaoLoginResponse;
import com.example.user.dto.KakaoTokenResponse;
import com.example.user.dto.KakaoUserInfoResponse;
import com.example.user.entity.User;
import com.example.user.repository.UserRepository;
import com.example.user.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final KakaoApiClient kakaoApiClient;
    private final UserRepository userRepository;
    private final JwtTokenProvider jwtTokenProvider;

    @Transactional
    public KakaoLoginResponse kakaoLogin(String authCode) {
        KakaoTokenResponse kakaoToken = kakaoApiClient.getAccessToken(authCode);
        KakaoUserInfoResponse userInfo = kakaoApiClient.getUserInfo(kakaoToken.getAccessToken());

        boolean isNewMember = false;
        User user = userRepository.findByKakaoId(userInfo.getId()).orElse(null);

        if (user == null) {
            isNewMember = true;
            user = User.builder()
                    .kakaoId(userInfo.getId())
                    .nickname(userInfo.getNickname())
                    .email(userInfo.getEmail())
                    .profileImageUrl(userInfo.getProfileImageUrl())
                    .build();
            user = userRepository.save(user);
        } else {
            user.updateProfile(
                    userInfo.getNickname(),
                    userInfo.getEmail(),
                    userInfo.getProfileImageUrl()
            );
        }

        String accessToken = jwtTokenProvider.generateAccessToken(user.getUuid().toString());
        String refreshToken = jwtTokenProvider.generateRefreshToken(user.getUuid().toString());

        return KakaoLoginResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .userId(user.getId())
                .nickname(user.getNickname())
                .isNewMember(isNewMember)
                .build();
    }
}

File: apps/user-service/src/auth/auth.service.ts
import { Injectable } from '@nestjs/common';
import { KakaoApiClient } from './kakao/kakao-api.client';
import { UserRepository } from '../user/user.repository';
import { JwtTokenProvider } from '../jwt/jwt-token.provider';
import { KakaoLoginResponseDto } from './dto/kakao-login-response.dto';
import { User } from '../user/user.entity';
import {
  getNickname,
  getEmail,
  getProfileImageUrl,
} from './kakao/kakao-user-info-response';

@Injectable()
export class AuthService {
  constructor(
    private readonly kakaoApiClient: KakaoApiClient,
    private readonly userRepository: UserRepository,
    private readonly jwtTokenProvider: JwtTokenProvider,
  ) {}

  async kakaoLogin(authCode: string): Promise<KakaoLoginResponseDto> {
    const kakaoToken = await this.kakaoApiClient.getAccessToken(authCode);
    const userInfo = await this.kakaoApiClient.getUserInfo(kakaoToken.access_token);

    let isNewMember = false;
    let user = await this.userRepository.findByKakaoId(String(userInfo.id));

    if (!user) {
      isNewMember = true;
      const newUser = new User();
      newUser.kakaoId = String(userInfo.id);
      newUser.nickname = getNickname(userInfo);
      newUser.email = getEmail(userInfo);
      newUser.profileImageUrl = getProfileImageUrl(userInfo);
      user = await this.userRepository.save(newUser);
    } else {
      user.updateProfile(
        getNickname(userInfo),
        getEmail(userInfo),
        getProfileImageUrl(userInfo),
      );
      user = await this.userRepository.save(user);
    }

    const accessToken = this.jwtTokenProvider.generateAccessToken(user.uuid);
    const refreshToken = this.jwtTokenProvider.generateRefreshToken(user.uuid);

    return {
      accessToken,
      refreshToken,
      userId: user.id,
      nickname: user.nickname,
      isNewMember,
    };
  }
}
